### PR TITLE
clusterd: unify storaged and computed

### PR DESCRIPTION
**Pull request by benesch**
Unify storaged and computed into a single binary named "clusterd". Both the storage controller and compute controller now create "clusterd" services, and simply ignore the half that they do not need.

The ultimate goal is to allow scheduling multiple sources, sinks, and indexes onto the same physical compute resources (i.e., a "cluster.") This PR is the next step on that path. See [0] for details.

[0]: https://www.notion.so/materialize/Cluster-unification-472efd2e9b8d4286bacc91b84d6b6f71

Touches https://github.com/MaterializeInc/cloud/issues/3642.

### Motivation

  * This PR takes a step towards a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Review by petrosagg — APPROVED** [review]
Changes look good!

File: src/clusterd/src/bin/clusterd.rs [hunk 1]
```diff
@@ -101,58 +117,42 @@ struct Args {
...
-        "storaged",
+        "clusterd",
         &args.tracing,
         mz_service::tracing::mz_sentry_event_filter,
         (BUILD_INFO.version, BUILD_INFO.sha, BUILD_INFO.time),
     )
     .await?;
-
-    // Keep this _after_ the mz_ore::tracing::configure call so that its panic
-    // hook runs _before_ the one that sends things to sentry.
```
Comment (teskje) [inline src/clusterd/src/bin/clusterd.rs:140]: What's up with this comment? Does it not apply anymore?

**Review by teskje — APPROVED** [review]
LGTM (modulo tests)

File: src/clusterd/src/bin/clusterd.rs [hunk 2]
```diff
@@ -195,38 +195,67 @@ async fn run(args: Args) -> Result<(), anyhow::Error> {
...
+            BUILD_INFO.semver_version(),
+            compute_client,
+            ProtoComputeServer::new,
+        ),
+    );
+
+    // TODO: unify storage and compute servers to use one timely cluster.
+
+    // Block forever.
+    future::pending().await
```
Comment (guswynn) [inline src/clusterd/src/bin/clusterd.rs:260]: i would just `select!` the 2 join handles here!

File: test/cluster/mzcompose.py [hunk 3]
```diff
@@ -743,7 +742,9 @@ def workflow_pg_snapshot_resumption(c: Composition) -> None:
         # Start postgres for the pg source
         Postgres(),
         Testdrive(no_reset=True),
-        Storaged(environment=["FAILPOINTS=pg_snapshot_failure=return"]),
+        Clusterd(
+            name="storaged", environment=["FAILPOINTS=pg_snapshot_failure=return"]
```
Comment (guswynn) [inline test/cluster/mzcompose.py:746]: might want a todo to clean this up!

**Review by guswynn — APPROVED** [review]

File: src/clusterd/src/bin/clusterd.rs (hunk 1 above)
Comment (benesch) [reply src/clusterd/src/bin/clusterd.rs:141]: It does! Just got lost in the refactor. Good eyes, thanks! Restored.

File: test/cluster/mzcompose.py (hunk 3 above)
Comment (benesch) [reply test/cluster/mzcompose.py]: Tests have all been adjusted to avoid using the terms "computed" or "storaged".

**Comment by benesch** [comment]
This is green, and nightly (finally) looks on track to be green, so 🚢. Thanks for the reviews, all!

File: src/clusterd/src/bin/clusterd.rs (hunk 2 above)
Comment (benesch) [reply src/clusterd/src/bin/clusterd.rs:272]: 🤷🏽 this matches what we do in `environmentd` (albeit in sync code), so just going to leave as is for now! Not opposed to cleaning it up in both places though.

**Pull request merged**